### DISCUSSION TITLE
What are forces?

### DISCUSSION THREAD

**Discussion by dvandersluis** [rubocop/rubocop#9255] [Q&A] [answered]
There isn't really any documentation about forces. What do forces do and why should a cop join one?

**Comment by bbatsov** [comment, marked as answer]
That's the best answer I can give is the commit that introduced them https://github.com/rubocop-hq/rubocop/commit/4b144e107ae6596abcba284f8af3ee6fa5d8b8fb

It seemed like an useful concept back then (and the name was a pun on "police force"), but we never fledged it out fully.

**Reply by marcandre** [reply]
> It seemed like an useful concept back then (and the name was a pun on "police force"), but we never fledged it out fully.

Indeed. And there's only one force today, to provide info about variable scope IIRC: `VariableForce`.

One limitation of forces in current implementation is that forces run separately from cops with a single callback. So a cop will get all the calls from the forces, then the callbacks from `on_send` etc. It's not possible to write a force currently that will trigger its callbacks among the normal callbacks. Because of this I believe all cops relying on `VariableForce` respond only to some of the force's callbacks and to no standard node callback.

**Reply by dvandersluis** [reply]
Thanks for the added info @marcandre!

**Comment by dvandersluis** [comment]
The puns are good :grin:

So I guess the idea was to run all the cops in a single force at once?

**Reply by bbatsov** [reply]
Yep, they were meant to be a performance optimization. The initial versions of RuboCop were pretty slow due to some very dumb things I had done originally (e.g. having each cop parse every file) and for a while we've had a lot of initiatives focused on improved performance.
